### PR TITLE
impr: add epo exclusion information (FPLA-1106)

### PR DESCRIPTION
**Pull request by lewisbirks**
### JIRA link (if applicable) ###

[FPLA-1106](https://tools.hmcts.net/jira/browse/FPLA-1106)

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOTypeValidator.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    @Override
+    public boolean isValid(Orders value, ConstraintValidatorContext context) {
+        if (featureToggleService.isEpoOrderTypeAndExclusionEnabled()) {
+            if (value.orderContainsEPO() && isNull(value.getEpoType())) {
+                return false;
+            }
+            return true;
+        }
+        return true;
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOTypeValidator.java:27]: or
```suggestion
        if (featureToggleService.isEpoOrderTypeAndExclusionEnabled()) {
            return !value.orderContainsEPO() || !isNull(value.getEpoType());
        }
        return true;
```
if you feel like the other is a bit too confusing

File: service/src/main/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEnteredEPOExcludedValidator.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    public boolean isValid(Orders value, ConstraintValidatorContext context) {
+        if (featureToggleService.isEpoOrderTypeAndExclusionEnabled()) {
+            if (value.orderContainsEPO() && directionsContainExclusionRequirement(value)
+                && isEmpty(value.getExcluded())) {
+                return false;
+            }
+            return true;
+        }
+        return true;
```
[the remark this answers is not in the document]
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEnteredEPOExcludedValidator.java:30]: or
```suggestion
        if (featureToggleService.isEpoOrderTypeAndExclusionEnabled()) {
            return !value.orderContainsEPO() || !directionsContainExclusionRequirement(value)
                || !isEmpty(value.getExcluded());
        }
        return true;
```
if you feel like the other is a bit too confusing

File: service/src/main/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOAddressValidator.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+public class HasEPOAddressValidator implements ConstraintValidator<HasEPOAddress, Orders> {
+
+    @Override
+    public boolean isValid(Orders value, ConstraintValidatorContext context) {
+        if (value.orderContainsEPO() && epoIsPreventRemoval(value)) {
+            if (isEmpty(value.getAddress().getAddressLine1()) || isEmpty(value.getAddress().getPostcode())) {
+                return false;
+            }
+        }
+        return true;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOAddressValidator.java:21]:
```suggestion
        if (value.orderContainsEPO() && epoIsPreventRemoval(value)) {
            return !isEmpty(value.getAddress().getAddressLine1()) && !isEmpty(value.getAddress().getPostcode());
        }
        return true;
```
or
```suggestion
        return !value.orderContainsEPO() || !epoIsPreventRemoval(value)
            || !isEmpty(value.getAddress().getAddressLine1())
            && !isEmpty(value.getAddress().getPostcode());
```

**Review by lewisbirks — COMMENTED** [review]
some simplifications to make sonar happy @toireasam

File: service/src/main/java/uk/gov/hmcts/reform/fpl/validation/interfaces/HasEnteredEPOExcluded.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.hmcts.reform.fpl.validation.interfaces;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/validation/interfaces/HasEnteredEPOExcluded.java:1]: minor: it would be nice to move these and the validators into their own epo package in here
i.e. `package uk.gov.hmcts.reform.fpl.validation.interfaces.epo;`

File: service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOAddressValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.gov.hmcts.reform.fpl.enums.EPOType.PREVENT_REMOVAL;
+import static uk.gov.hmcts.reform.fpl.enums.EPOType.REMOVE_TO_ACCOMMODATION;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = {JacksonAutoConfiguration.class, FeatureToggleService.class})
+class HasEPOAddressValidatorTest extends AbstractValidationTest {
+
+    @MockBean
+    private FeatureToggleService featureToggleService;
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOAddressValidatorTest.java:26]: If you don't want to have to define a redundant feature toggle service one way of doing it would be to make an interface with the required methods that has the validation annotation on it, the orders object can then implment that interface and the validator could validate against that interface. The test can then create a new object with implementing that interface, and then you would only be testing against that annotation and not having to accomodate for all the others on the object. Thinking about it, this might be too much work for just removing a redundant bean but food for thought

**Review by lewisbirks — COMMENTED** [review]
Looks good, some vv minor comments, will have another pass over in a bit

**Comment by lewisbirks** [comment]
Might be worth adding more tests to the validators that haven't got conditions covered in the code coverage, not too fussed about the docmosis stuff but the validators sole purpose is based around those conditions so all flows should be considered

File: service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOAddressValidatorTest.java (hunk 5 above)
Comment (toireasam) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOAddressValidatorTest.java:26]: really good idea, might be a bit too much work considering we will be removing this dependency shortly as it's a toggle. will leave for now as may require a fair bit of time to get it implemented.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/docmosis/CaseSubmissionGenerationService.java [hunk 6]
```diff
@@ -187,6 +188,15 @@ private void appendOtherOrderToOrdersNeeded(final Orders orders, final StringBui
     private void appendEmergencyProtectionOrdersAndDetailsToOrdersNeeded(final Orders orders,
                                                                          final StringBuilder stringBuilder) {
         if (isNotEmpty(orders.getEmergencyProtectionOrders())) {
+            if (isNotEmpty(orders.getEpoType())) {
+                stringBuilder.append(orders.getEpoType().getLabel());
+                stringBuilder.append(NEW_LINE);
+                if (orders.getEpoType() == PREVENT_REMOVAL) {
+                    String address = orders.getAddress().getAddressAsString("\n");
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/docmosis/CaseSubmissionGenerationService.java:195]: Could use the `NEW_LINE` constant instead of passing `"\n"`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOAddressValidator.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+import static org.apache.logging.log4j.util.Strings.isEmpty;
+import static uk.gov.hmcts.reform.fpl.enums.EPOType.PREVENT_REMOVAL;
+
+public class HasEPOAddressValidator implements ConstraintValidator<HasEPOAddress, Orders> {
+
+    @Override
+    public boolean isValid(Orders value, ConstraintValidatorContext context) {
+        if (value.orderContainsEPO() && epoIsPreventRemoval(value)) {
+            return !isEmpty(value.getAddress().getAddressLine1()) && !isEmpty(value.getAddress().getPostcode());
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEPOAddressValidator.java:17]: vv minor but it might be nice to extract this internal logic into a method something like `hasPopulatedAddress(value.getAddress())`

File: service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEnteredEPOExcludedValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import uk.gov.hmcts.reform.fpl.enums.OrderType;
+import uk.gov.hmcts.reform.fpl.model.Orders;
+import uk.gov.hmcts.reform.fpl.service.FeatureToggleService;
+import uk.gov.hmcts.reform.fpl.validation.AbstractValidationTest;
+
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.BDDMockito.given;
+import static uk.gov.hmcts.reform.fpl.config.utils.EmergencyProtectionOrderDirectionsType.EXCLUSION_REQUIREMENT;
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEnteredEPOExcludedValidatorTest.java:18]: probs best done in a different PR but these enums shouldn't be in the config package

Comment (toireasam) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/HasEnteredEPOExcludedValidatorTest.java:18]: will keep in seperate pr as not to inflate this one

**Review by toireasam — DISMISSED** [review]
Lewis is happy with this one, my pr which he reviewed. so approving.

**Review by toireasam — APPROVED** [review]
